distributed/recreate_exceptions.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import print_function, division, absolute_import
+
+import logging
+from tornado import gen
+from .client import futures_of
+from .utils import sync
+from .utils_comm import pack_data
+from .worker import _deserialize
+
+logger = logging.getLogger(__name__)
+
+
+class ReplayExceptionScheduler(object):
+    """ A plugin for the scheduler to recreate exceptions locally
+
+    This adds the following routes to the scheduler
+
+    *  cause_of_failure
+    """
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.scheduler.handlers['cause_of_failure'] = self.cause_of_failure
+        self.scheduler.extensions['exceptions'] = self
+
+    def cause_of_failure(self, *args, **kwargs):
+        """
+        Return details of first failed task required by set of keys
+
+        Parameters
+        ----------
+        keys: list of keys known to the scheduler
+
+        Returns
+        -------
+        Dictionary with:
+        cause: the key that failed
+        task: the definition of that key
+        deps: keys that the task depends on
+        """
+
+        keys = kwargs.pop('keys', [])
+        for key in keys:
+            if isinstance(key, list):
+                key = tuple(key)
+                key = tuple(key)
```

Comment:
Perhaps drop the check and just call 

    key = tuple(key)  # ensure not a list from msgpack